File: Android/app/src/main/java/com/hhp227/datemate/data/PostRepository.kt
package com.hhp227.datemate.data

import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.database.DataSnapshot
import com.google.firebase.database.DatabaseError
import com.google.firebase.database.DatabaseReference
import com.google.firebase.database.FirebaseDatabase
import com.google.firebase.database.ValueEventListener
import com.hhp227.datemate.data.Post
import kotlinx.coroutines.channels.awaitClose
import kotlinx.coroutines.channels.trySendBlocking
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.callbackFlow

class PostRepository {
    private val rootRef: DatabaseReference = FirebaseDatabase.getInstance().reference

    private val postRef: DatabaseReference = rootRef.child("posts")

    private val userPostRef: DatabaseReference = rootRef.child("user-posts")

    fun getPosts(): Flow<List<Post>> = callbackFlow {
        val postListener = object : ValueEventListener {
            override fun onDataChange(snapshot: DataSnapshot) {
                trySend(snapshot.children.mapNotNull { dataSnapshot ->
                    dataSnapshot.key?.let { dataSnapshot.getValue(Post::class.java)?.apply { key = it } } ?: Post()
                })
            }

            override fun onCancelled(error: DatabaseError) {
                trySend(emptyList())
            }
        }

        postRef.addListenerForSingleValueEvent(postListener)
        awaitClose {
            postRef.removeEventListener(postListener)
            close()
        }
    }

    fun getPost(key: String) = callbackFlow {
        postRef.child(key)
            .get()
            .addOnCompleteListener { task ->
                if (task.isSuccessful) {
                    trySendBlocking(task.result.key?.let {
                        task.result.getValue(Post::class.java)?.apply { this.key = it }
                    } ?: Post())
                }
            }.addOnFailureListener {
                trySendBlocking(Post())
            }
        awaitClose { close() }
    }

    fun addPost(title: String, content: String) {

    }

    fun removePost(key: String) {

    }

    fun getUserPostKeys(): Flow<List<String>> {
        val mutableStateFlow = MutableStateFlow<List<String>>(emptyList())

        FirebaseAuth.getInstance().currentUser?.also { user ->
            userPostRef.child(user.uid).get().addOnCompleteListener { task ->
                if (task.isSuccessful) {
                    mutableStateFlow.value = task.result.children.mapNotNull { it.key }
                }
            }
        }
        return mutableStateFlow
    }
}

File: iOS/DateMate/Data/PostRepository.swift
//
//  PostDetailRepository.swift
//  DateMate
//
//  Created by 홍희표 on 2021/12/14.
//

import Foundation
import FirebaseDatabase
import Combine
import FirebaseAuth

class PostRepository {
    private let rootRef: DatabaseReference
    
    private let postRef: DatabaseReference
    
    private let userPostRef: DatabaseReference

    func getPost(_ key: String) -> AnyPublisher<Post, Error> {
        return postRef.child(key).observeSingleEvent(of: .value).tryMap { result in
            if let dic = result.value as? [String: Any] {
                return Post(
                    id: dic["uid"] as! String,
                    author: dic["author"] as! String,
                    title: dic["title"] as! String,
                    body: dic["body"] as! String,
                    starCount: dic["starCount"] as! Int,
                    stars: [:],
                    key: result.key
                )
            } else {
                fatalError()
            }
        }.eraseToAnyPublisher()
    }
    
    func addPost(_ title: String, _ content: String) -> AnyPublisher<DatabaseReference, Error> {
        guard let user = Auth.auth().currentUser, let key = postRef.childByAutoId().key else {
            fatalError()
        }
        let postValue: [String: Any] = [
            "uid": user.uid,
            "author": user.email?.split(separator: "@").first ?? "author",
            "title": title,
            "body": content,
            "starCount": 0,
            "stars": [:]
        ]
        let childUpdates: [String: Any] = [
            "/posts/\(key)": postValue,
            "/user-posts/\(user.uid)/\(key)": postValue
        ]
        return rootRef.updateChildValues(childUpdates).eraseToAnyPublisher()
    }
    
    func removePost(_ key: String) -> AnyPublisher<DatabaseReference, Error> {
        guard let user = Auth.auth().currentUser else { fatalError() }
        let childUpdates: [String: Any?] = [
            "/posts/\(key)": nil,
            "/user-posts/\(user.uid)/\(key)": nil
        ]
        return rootRef.updateChildValues(childUpdates).eraseToAnyPublisher()
    }
    
    func getUserPostKeys(_ key: String) -> AnyPublisher<[String], Error> {
        guard let user = Auth.auth().currentUser else { fatalError() }
        return userPostRef.child(user.uid).observer(for: .value).tryMap { dataSnapshot in
            dataSnapshot.children.map { ($0 as! DataSnapshot).key }
        }.eraseToAnyPublisher()
    }
    
    init() {
        self.rootRef = Database.database().reference()
        self.postRef = rootRef.child("posts")
        self.userPostRef = rootRef.child("user-posts")
    }
}
